[composable] Enable replicate + trec_shard overall

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #98890
* #98889

replicate + trec_shard works if we shard / replicate individually, such as follows:

```
m = TestSparseNN()
shard(m.sparse)
replicate(m.dense)
```

but does not work if users do the following:
```
m = TestSparseNN()
shard(m, sharders=[...])
replicate(m)
```

Many upstream trainers use the latter use case, as sharding is not done on individual module level but rather overall module by specifying planners that contain logic for how to shard different embedding table types.

This diff enables the latter approach (while keeping the former intact), but users need to specify `ignored_modules` to ignore embedding tables in replicate(). This is similar to FSDP (class based and composable) and DDP today.

Differential Revision: [D44899155](https://our.internmc.facebook.com/intern/diff/D44899155/)

rohan-varma committed:
[composable] Enable replicate + trec_shard overall

replicate + trec_shard works if we shard / replicate individually, such as follows:

```
m = TestSparseNN()
shard(m.sparse)
replicate(m.dense)
```

but does not work if users do the following:
```
m = TestSparseNN()
shard(m, sharders=[...])
replicate(m)
```

Many upstream trainers use the latter use case, as sharding is not done on individual module level but rather overall module by specifying planners that contain logic for how to shard different embedding table types.

This diff enables the latter approach (while keeping the former intact), but users need to specify `ignored_modules` to ignore embedding tables in replicate(). This is similar to FSDP (class based and composable) and DDP today.

Differential Revision: [D44899155](https://our.internmc.facebook.com/intern/diff/D44899155/)

[ghstack-poisoned]

mrshenli commented on torch/distributed/_composable/replicate.py:
```diff
@@ -23,62 +25,49 @@ def replicate(
+        )
 
     def mark_module(self, module: nn.Module, **kwargs) -> None:
-        if not _can_compose(module):
```
any reason for removing this?

We will remove this eventually, but curious about does this prevent integrating with trec_shard?

rohan-varma replied:
Actually we can keep this, because _can_compose only checked whether replicate can wrap a fully_shard which is unsupported at the moment.

mrshenli commented on torch/distributed/_composable/replicate.py:
```diff
@@ -23,62 +25,49 @@ def replicate(
 
-    def _recursive_collect_params(self, module: nn.Module) -> None:
-        # skip if managed by other APIs
-        if not _can_compose(module):
```
when, without this, will it still work if a submodule is wrapped by `fully_shard`?

mrshenli commented on torch/distributed/_composable/replicate.py:
```diff
@@ -1,15 +1,17 @@
 @contract()
 def replicate(
     module: nn.Module,  # NOTE: contract now supports single module only
+    ignored_modules: Optional[Iterable[torch.nn.Module]] = None,
```
curious, can trec_shard add entry to registry on submodules it swapped, so that replicate can detect automatically?

yhcharles replied:
Agree, and that's the spirit of these *composable* APIs, `ignore_modules` may look like a temporary workaround solution.

rohan-varma replied:
This is indeed the right way to do so. currently trec_shard adds an entry for the top level, but does not do so for individually sharded submodules. 

This is also why fully_shard needs ignored_modules today. For this PR, I think it'd be best to unblock by adding ignored_module support, and then follow up with trec_shard to fix the registry and remove ignored_modules.

Also, if user calls replicate() first and then trec_shard, we would still need `ignored_modules`, right? Unless we add a policy or similar to replicate API.

yhcharles commented on torch/distributed/_composable/replicate.py:
```diff
@@ -23,62 +25,49 @@ def replicate(
+            return  # if module A is ignored, all of A's children are also ignored.
+
+        ignored_params: Set[torch.nn.Parameter] = {
+            p for m in self.ignored_modules for p in m.parameters()
```
The original idea to make `replicate` to be composable is: assuming we use it on a given Module, then it will recursively check all the sub-modules, apply data parallel on the sub-modules only when the sub-module is not marked by another composable API (since all composable APIs follows `contract`, they are supposed to be able to know each other).
It seems this change removes this logic, and makes `replicate` not composable. Is there any reason or it's my misunderstanding?
cc @mrshenli

rohan-varma replied:
IIUC this was not completely the previous behavior.

Previously, `_can_compose` only checked for fully_shard, so trec_shard would get through. Thus when I tried to do a top-level replicate, replicate would try to mark embedding modules as data parallel which is not desired. 

The right fix would be to add the registry check for trec_shard as exists for _can_compose. However, trec_shard does not seem to add to the registry when it is wrapping an embedding submodule (only adds it for the top level). Once we fix this in trec_shard we may be able to get rid of ignore_modules. 

We should also make sure fully_shard can get rid of it as well then.

yhcharles replied:
Yes, the previous implementation was not complete -- only `fully_shard` and `replicate` run the check and thus are composable with each other. And also agree that the right fix is to enable the check for trec_shard.

Q:
- If we plan to add this check to trec_shard, then why removing it from `replicate`? It seems adding `ignored_modules` has not conflicts with the check.
- Will removing this check break the existing composability of `fully_shard` and `replicate`?

rohan-varma committed:
Update on "[composable] Enable replicate + trec_shard overall"


replicate + trec_shard works if we shard / replicate individually, such as follows:

```
m = TestSparseNN()
shard(m.sparse)
replicate(m.dense)
```

but does not work if users do the following:
```
m = TestSparseNN()
shard(m, sharders=[...])
replicate(m)
```

Many upstream trainers use the latter use case, as sharding is not done on individual module level but rather overall module by specifying planners that contain logic for how to shard different embedding table types.

This diff enables the latter approach (while keeping the former intact), but users need to specify `ignored_modules` to ignore embedding tables in replicate(). This is similar to FSDP (class based and composable) and DDP today.

Differential Revision: [D44899155](https://our.internmc.facebook.com/intern/diff/D44899155/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[composable] Enable replicate + trec_shard overall"


replicate + trec_shard works if we shard / replicate individually, such as follows:

```
m = TestSparseNN()
shard(m.sparse)
replicate(m.dense)
```

but does not work if users do the following:
```
m = TestSparseNN()
shard(m, sharders=[...])
replicate(m)
```

Many upstream trainers use the latter use case, as sharding is not done on individual module level but rather overall module by specifying planners that contain logic for how to shard different embedding table types.

This diff enables the latter approach (while keeping the former intact), but users need to specify `ignored_modules` to ignore embedding tables in replicate(). This is similar to FSDP (class based and composable) and DDP today.

Differential Revision: [D44899155](https://our.internmc.facebook.com/intern/diff/D44899155/)

[ghstack-poisoned]

mrshenli commented on test/distributed/_composable/test_replicate.py:
```diff
@@ -136,6 +133,38 @@ def test_replicate_single_module(self):
         replicate_model = replicate(deepcopy(model))
         self._compare_module(model, replicate_model)
 
+    @skip_if_lt_x_gpu(2)
```
default world size in `MultiProcessTestCase` is 4. We will need to override `world_size` property or change this to 4 as well.

```python
@property
    def world_size(self) -> int:
      return 2
```

mrshenli reviewed (approved):
stamp to unblock, please fix lint before landing.

rohan-varma committed:
Update on "[composable] Enable replicate + trec_shard overall"


replicate + trec_shard works if we shard / replicate individually, such as follows:

```
m = TestSparseNN()
shard(m.sparse)
replicate(m.dense)
```

but does not work if users do the following:
```
m = TestSparseNN()
shard(m, sharders=[...])
replicate(m)
```

Many upstream trainers use the latter use case, as sharding is not done on individual module level but rather overall module by specifying planners that contain logic for how to shard different embedding table types.

This diff enables the latter approach (while keeping the former intact), but users need to specify `ignored_modules` to ignore embedding tables in replicate(). This is similar to FSDP (class based and composable) and DDP today.

Differential Revision: [D44899155](https://our.internmc.facebook.com/intern/diff/D44899155/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[composable] Enable replicate + trec_shard overall"


replicate + trec_shard works if we shard / replicate individually, such as follows:

```
m = TestSparseNN()
shard(m.sparse)
replicate(m.dense)
```

but does not work if users do the following:
```
m = TestSparseNN()
shard(m, sharders=[...])
replicate(m)
```

Many upstream trainers use the latter use case, as sharding is not done on individual module level but rather overall module by specifying planners that contain logic for how to shard different embedding table types.

This diff enables the latter approach (while keeping the former intact), but users need to specify `ignored_modules` to ignore embedding tables in replicate(). This is similar to FSDP (class based and composable) and DDP today.

Differential Revision: [D44899155](https://our.internmc.facebook.com/intern/diff/D44899155/)

[ghstack-poisoned]

yhcharles commented on torch/distributed/_composable/replicate.py:
```diff
@@ -52,34 +61,30 @@ def mark_module(self, module: nn.Module, **kwargs) -> None:
-        for child in module.children():
-            self._recursive_collect_params(child)
+
+        for child_module in module.children():
+            self._collect_params(child_module)
```
Thanks, LGTM

rohan-varma committed:
Update on "[composable] Enable replicate + trec_shard overall"


replicate + trec_shard works if we shard / replicate individually, such as follows:

```
m = TestSparseNN()
shard(m.sparse)
replicate(m.dense)
```

but does not work if users do the following:
```
m = TestSparseNN()
shard(m, sharders=[...])
replicate(m)
```

Many upstream trainers use the latter use case, as sharding is not done on individual module level but rather overall module by specifying planners that contain logic for how to shard different embedding table types.

This diff enables the latter approach (while keeping the former intact), but users need to specify `ignored_modules` to ignore embedding tables in replicate(). This is similar to FSDP (class based and composable) and DDP today.

Differential Revision: [D44899155](https://our.internmc.facebook.com/intern/diff/D44899155/)

[ghstack-poisoned]

rohan-varma commented:
@pytorchbot merge -g

rohan-varma commented:
@pytorchbot merge
